### ISSUE TITLE
`type foo = u8` should create a new type, not an alias

### ISSUE THREAD

**Issue by sbillig** [ethereum/fe#556]
@g-r-a-n-t sort of suggested this, in the context of defining built-in numeric types in fe code, in terms of u256.
```
type u8 = u256 # instead of `struct u8: value u256` or `struct u8(u256)`
type u16 = u256
impl Add for u8: ...
impl Add<u16> for u8: ...
```

I think in general, creating a new type instead of an alias for the underlying type would be more useful for writing correct code. (Note that these types would only exist at compile time, and don't add any overhead at runtime. The generated code would just use the underlying type.)
```
type Gwei = u256
type Timestamp = u256

fn transfer(to: address, at: Timestamp, amount: Gwei):
  ...

fn send_some_money(to: address):
  let time = Timestamp(block.timestamp)
  let amount = Gwei(100)
  transfer(to, amount, time)  # whoops!
```

Is there a use-case where a simple type alias would be preferred? I haven't spent a lot of time thinking about this yet, but let's discuss.

Alternatives:
- "tuple structs", as in rust. Eg. `struct Gwei(u256)`
I think this would be a fine solution, but we'll have to allow structs (of size <= 32) to be value types somehow (which we should probably do anyway), and make sure there's no runtime overhead.
- rely on function argument labels to provide some of this protection, perhaps by making argument labels required (#513)

**Comment by cburgdorf** [comment]
I think the main use case for type aliases is to cleanup function signatures eg. to avoid something like:

```
pub fn foo (self) -> Array<Array<ProblemFactory<Problem>>>
   ...
```

In that case a type alias can help:

```
type UsefulName = Array<ProblemFactory<Problem>>
pub fn foo (self) -> Array<UsefulName>
   ...
```

Here `Array<UsefulName>` and `Array<Array<ProblemFactory<Problem>>>` are fully interchangeable but the shorter name is easier for the eye.

**Comment by cburgdorf** [comment]
>In that example, would it be inconvenient if `UsefulName` was a new type rather than just an alias?

Yes, I think this might be inconvenient in certain cases. It may trigger a cascade of changes when all you wanted was to cleanup a few function signatures in a single file.

>Maybe we should just copy rust and have type aliases and tuple structs

Yes, that's what I'm thinking. Both serve different purposes and  tuple structs can serve well where one wants to create a new type that can not be used interchangeably (e.g. to avoid mixing *miles* and *kilometer*):

```
struct Miles(u256)
struct Kilometers(u256)

pub fn can_only_use_miles(distance: Miles):
   pass

pub fn can_only_use_km(distance: Kilometers):
   pass
```
